### PR TITLE
[cmd] Add a `arch` command for managing `gef.arch`

### PR DESCRIPTION
**Pull request by ValekoZ**
Available subcommands:
    - get  -> prints the current arch and the reason
    - set  -> manually set the current arch (or auto-detect if no arg is provided)
    - list -> list available architectures

Implements this: https://github.com/hugsy/gef/issues/1002

### CODE REVIEW

File: gef.py [hunk 1]
```diff
@@ -4757,6 +4761,62 @@ def __set_repeat_count(self, argv: List[str], from_tty: bool) -> None:
         return
 
 
+@register
+class PieCommand(GenericCommand):
+    """Manage the current loaded architecture"""
```
Comment (Grazfather) [inline gef.py:4766]: nit: Period at the end for rendered docs.

File: gef.py [hunk 2]
```diff
@@ -4757,6 +4761,62 @@ def __set_repeat_count(self, argv: List[str], from_tty: bool) -> None:
...
+
+    _cmdline_ = "arch"
+    _syntax_ = f"{_cmdline_} (list|get|set) ..."
+    _example_ = f"{_cmdline_}"
```
Comment (Grazfather) [inline gef.py:4770]: better example?

File: gef.py [hunk 3]
```diff
@@ -4757,6 +4761,62 @@ def __set_repeat_count(self, argv: List[str], from_tty: bool) -> None:
         return
 
 
+@register
+class PieCommand(GenericCommand):
```
Comment (Grazfather) [inline gef.py:4765]: Piecommand?

File: gef.py [hunk 4]
```diff
@@ -4757,6 +4761,62 @@ def __set_repeat_count(self, argv: List[str], from_tty: bool) -> None:
...
+        gef_print(f"{Color.greenify('Reason')}: {gef.arch_reason}")
+
+
+@register
+class ArchSetCommand(GenericCommand):
+    """Set the current loaded architecture"""
+
+    _cmdline_ = "arch set"
+    _syntax_ = f"{_cmdline_} <arch>"
+    _example_ = f"{_cmdline_}"
```
Comment (Grazfather) [inline gef.py:4800]: better example? e.g. arch set x86

File: gef.py [hunk 5]
```diff
@@ -4757,6 +4761,62 @@ def __set_repeat_count(self, argv: List[str], from_tty: bool) -> None:
...
+
+    def do_invoke(self, args: List[str]) -> None:
+        gef_print(f"{Color.greenify('Arch')}: {gef.arch}")
+        gef_print(f"{Color.greenify('Reason')}: {gef.arch_reason}")
+
+
+@register
+class ArchSetCommand(GenericCommand):
+    """Set the current loaded architecture"""
```
Comment (Grazfather) [inline gef.py:4796]: periods in doc strings

File: gef.py [hunk 6]
```diff
@@ -4757,6 +4761,62 @@ def __set_repeat_count(self, argv: List[str], from_tty: bool) -> None:
...
+    """List the available architectures"""
+
+    _cmdline_ = "arch list"
+    _syntax_ = f"{_cmdline_}"
+    _example_ = f"{_cmdline_}"
+
+    def do_invoke(self, args: List[str]) -> None:
+        gef_print(Color.greenify("Available architectures:"))
+        for arch in __registered_architectures__.keys():
+            if type(arch) == str:
```
Comment (Grazfather) [inline gef.py:4816]: Why would the registered architectures not be keyed by strings?

File: gef.py (hunk 1 above)
Comment (ValekoZ) [reply gef.py]: Done

File: gef.py (hunk 2 above)
Comment (ValekoZ) [reply gef.py]: Done

File: gef.py (hunk 3 above)
Comment (ValekoZ) [reply gef.py]: Ups :sweat_smile: 
Done

File: gef.py (hunk 4 above)
Comment (ValekoZ) [reply gef.py]: Done

File: gef.py (hunk 5 above)
Comment (ValekoZ) [reply gef.py]: Done

File: gef.py (hunk 6 above)
Comment (ValekoZ) [reply gef.py:4816]: Because of Abi objects that can't be used as keys from the `arch set` command

**Review by hugsy — CHANGES_REQUESTED** [review]
All good except that it's missing tests

**Review by hugsy — CHANGES_REQUESTED** [review]
minor stuff to tweak and we're good

File: gef.py [hunk 7]
```diff
@@ -4757,6 +4761,62 @@ def __set_repeat_count(self, argv: List[str], from_tty: bool) -> None:
...
+@register
+class ArchSetCommand(GenericCommand):
+    """Set the current loaded architecture."""
+
+    _cmdline_ = "arch set"
+    _syntax_ = f"{_cmdline_} <arch>"
+    _example_ = f"{_cmdline_} X86"
+
+    def do_invoke(self, args: List[str]) -> None:
+        reset_architecture(args[0] if args else None)
```
[the remark this answers is not in the document]
Comment (ValekoZ) [reply gef.py:4804]: Does `reset_architecture` should be case insensitive? Or just the command?

File: gef.py [hunk 8]
```diff
@@ -4758,6 +4767,68 @@ def __set_repeat_count(self, argv: List[str], from_tty: bool) -> None:
...
+    _cmdline_ = "arch list"
+    _syntax_ = f"{_cmdline_}"
+    _example_ = f"{_cmdline_}"
+
+    def do_invoke(self, args: List[str]) -> None:
+        gef_print(Color.greenify("Available architectures:"))
+        for arch in sorted(set(__registered_architectures__.values()), key=lambda x: x.arch):
+            if arch != GenericArchitecture:
+                gef_print(' ' + Color.yellowify(str(arch())))
+                for alias in filter(lambda x: isinstance(x, str), arch.aliases):
```
Comment (Grazfather) [inline gef.py:4828]: slight preference for comprehensions here e.g. `for alias in (x for x in arch.aliases if isinstance(x, str))`

File: gef.py [hunk 9]
```diff
@@ -11554,6 +11625,7 @@ class Gef:
     def __init__(self) -> None:
         self.binary: Optional[FileFormat] = None
         self.arch: Architecture = GenericArchitecture() # see PR #516, will be reset by `new_objfile_handler`
+        self.arch_reason: str = "This default architecture"
```
Comment (Grazfather) [inline gef.py:11628]: The

File: gef.py [hunk 10]
```diff
@@ -4758,6 +4767,68 @@ def __set_repeat_count(self, argv: List[str], from_tty: bool) -> None:
...
+    """List the available architectures."""
+
+    _cmdline_ = "arch list"
+    _syntax_ = f"{_cmdline_}"
+    _example_ = f"{_cmdline_}"
+
+    def do_invoke(self, args: List[str]) -> None:
+        gef_print(Color.greenify("Available architectures:"))
+        for arch in sorted(set(__registered_architectures__.values()), key=lambda x: x.arch):
+            if arch != GenericArchitecture:
```
Comment (Grazfather) [inline gef.py:4826]: Cleaner to do
```python
if arch is Generic Architecture:
    continue
```

File: gef.py (hunk 8 above)
Comment (Grazfather) [reply gef.py:4828]: or maybe easiest would be
```python
for alias in arch.aliases:
    if isinstance(alias, str):
        gef_print(f"  {alias}")
```

One liners are nicer but not when they make things more confusing.

File: tests/commands/arch.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+Arch commands test module
+"""
+
+from tests.base import RemoteGefUnitTestGeneric
+from tests.utils import ARCH
+import pytest
+
+
+class ArchCommand(RemoteGefUnitTestGeneric):
+    """Generic class for command testing, that defines all helpers"""
```
Comment (Grazfather) [inline tests/commands/arch.py:11]: Period

File: tests/commands/arch.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
+"""
+Arch commands test module
+"""
+
+from tests.base import RemoteGefUnitTestGeneric
+from tests.utils import ARCH
+import pytest
```
Comment (Grazfather) [inline tests/commands/arch.py:7]: put third party libs in a separate paragraph before our libs.
stdlib
blank
pytest
blank
ours

File: tests/commands/arch.py (hunk 11 above)
Comment (Grazfather) [reply tests/commands/arch.py:11]: Is this accurate?

File: tests/commands/arch.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+
+class ArchCommand(RemoteGefUnitTestGeneric):
+    """Generic class for command testing, that defines all helpers"""
+
+    def setUp(self) -> None:
```
Comment (Grazfather) [inline tests/commands/arch.py:13]: Is this required?

File: tests/commands/arch.py [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+        return super().setUp()
+
+    @pytest.mark.skipif(ARCH != "x86_64", reason=f"Skipped for {ARCH}")
+    def test_cmd_arch_get(self):
+        gdb = self._gdb
+
+        res = gdb.execute("arch get", to_string=True)
+        self.assertIn(" Architecture(X86, 64, LITTLE_ENDIAN)", res)
```
Comment (Grazfather) [inline tests/commands/arch.py:21]: we can do this later (I am sure many tests do this 'wrong' but with the pytest runner we get better errors logs if you do
`assert "..." in res`
and you can even explain
`assert "..." in res, "Arch should be x86 64"`

File: tests/commands/arch.py (hunk 11 above)
Comment (ValekoZ) [reply tests/commands/arch.py:11]: Ups copy pasted blindly ^^

File: gef.py [hunk 15]
```diff
@@ -4758,6 +4767,71 @@ def __set_repeat_count(self, argv: List[str], from_tty: bool) -> None:
...
+    """Set the current loaded architecture."""
+
+    _cmdline_ = "arch set"
+    _syntax_ = f"{_cmdline_} <arch>"
+    _example_ = f"{_cmdline_} X86"
+
+    def do_invoke(self, args: List[str]) -> None:
+        reset_architecture(args[0].lower() if args else None)
+
+    def complete(self, text: str, word: str) -> List[str]:
```
Comment (hugsy) [inline gef.py:4811]: :tada:

File: tests/commands/arch.py [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+        gdb = self._gdb
+
+        gdb.execute("arch set X86")
+
+        res = gdb.execute("arch get", to_string=True)
+        assert " Architecture(X86, 32, LITTLE_ENDIAN)" in res
+        assert " The architecture has been set manually" in res
+
+
+        gdb.execute("arch set ppc")
```
Comment (hugsy) [inline tests/commands/arch.py:32]: awesome

**Review by hugsy — APPROVED** [review]

**Pull request merged**